clash-cores/src/Clash/Cores/UART.hs

```diff
@@ -0,0 +1,205 @@
+{-|
+  Copyright   :  (C) 2021, LUMI GUIDE FIETSDETECTIE B.V.
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  Rowan Goemans <goemansrowan@gmail.com>
+
+  UART transmitter and receiver module
+-}
+
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE NumericUnderscores #-}
+
+module Clash.Cores.UART(HzToPeriod, uartTx, uartRx, uart) where
+
+import Clash.Prelude
+
+-- | Convert frequency in hertz to period in picoseconds
+type HzToPeriod (rate :: Nat) = Div 1_000_000_000_000 rate
+
+-- | State used for the UART transmitter
+data UartTxState (period :: Nat)
+   -- | Not transmitting anything
+   = UartTxIdle
+   -- | Transmitting a word
+   | UartTxTransmitting
+       { uartTxWord :: BitVector 10
+       -- ^ All databits that need to be transmitted
+       -- Includes start and stop bit
+       , uartTxPeriodCounter :: Index period
+       -- ^ Counter that tracks how many cycles we need to hold the
+       -- current bit on the tx line
+       , uartTxWordCounter :: Index 10
+       -- ^ Which bit we are transmitting
+       }
+  deriving (Show, Generic, NFDataX)
+
+-- | Create a word that includes start a stop bits from a raw word
+createTxWord
+  :: BitVector 8
+  -- ^ The word to transmit
+  -> BitVector 10
+  -- ^ The word including start and stop bit
+createTxWord d = (1 :: (BitVector 1)) ++# d ++# (0 :: (BitVector 1))
+
+-- | State transition function for the UART transmitter
+uartTxT
+  :: forall (period :: Nat)
+   . KnownNat period
+  => UartTxState period
+  -- ^ The current state
+  -> Maybe (BitVector 8)
+  -- ^ The word to transmit
+  -> (UartTxState period, (Bit, Bool))
+  -- ^ The next state, the output bit to the tx line and an acknowledgement
+  -- that the word to transmit has been received
+uartTxT UartTxIdle (Just d) = (nextState, (1, True))
+  where
+    nextState = UartTxTransmitting
+                  { uartTxWord = createTxWord d
+                  , uartTxPeriodCounter = 0
+                  , uartTxWordCounter = 0
+                  }
+uartTxT UartTxIdle Nothing = (UartTxIdle, (1, False))
+uartTxT state@UartTxTransmitting{..} _ = (nextState, (lsb uartTxWord, False))
+  where
+    periodDone = uartTxPeriodCounter == maxBound
+    lastBit = uartTxWordCounter == maxBound
```

Comment:
`uartTxWordCounter` is superfluous: since the last bit in `uartTxWord` is always a one (stop bit), this suffices:
```suggestion
    lastBit = uartTxWord' == repeat 0
```
when
```
  uartTxWord' = 0 +>> uartTxWord
```
(this is equal to the record update a few lines down)

Reply:
So this comment is slightly outdated due to the update from `uartTxWord :: Vec 10 Bit` to `uartTxWord :: BitVector 10`.

```suggestion
    lastBit = uartTxWord' == 0
```

```
  uartTxWord' = shiftR uartTxWord 1
```

Reply:
```suggestion
    lastBit = uartTxWord == 1
```
if you'd rather match on current state. Sometimes you overlook the obvious :-).